File: libs/python/cua-cli/cua_cli/commands/image.py
"""Image management commands for CUA CLI.

Handles both cloud images (push/pull to CUA cloud) and local images
(create, clone, shell, info).
"""

import argparse
from datetime import datetime
from pathlib import Path
from typing import Any

import aiohttp
from cua_cli.api.client import CloudAPIClient, calculate_file_hash, format_bytes
from cua_cli.utils.async_utils import run_async
from cua_cli.utils.output import (
    print_error,
    print_info,
    print_json,
    print_success,
    print_table,
)

# Default local image storage
LOCAL_IMAGES_DIR = Path.home() / ".local" / "share" / "cua" / "images"

# Default part size for multi-part upload: 100MB
DEFAULT_PART_SIZE = 100 * 1024 * 1024


def register_parser(subparsers: argparse._SubParsersAction) -> None:
    """Register the image command and subcommands."""
    for cmd_name in ("image", "img"):
        img_parser = subparsers.add_parser(
            cmd_name,
            help="Image management commands",
            description="Manage VM images (cloud and local)",
        )

        img_subparsers = img_parser.add_subparsers(
            dest="image_command",
            help="Image command",
        )

        # list command
        list_parser = img_subparsers.add_parser(
            "list",
            aliases=["ls"],
            help="List images",
        )
        list_parser.add_argument(
            "--json",
            action="store_true",
            help="Output as JSON",
        )
        list_parser.add_argument(
            "--local",
            action="store_true",
            help="List local images only",
        )
        list_parser.add_argument(
            "--cloud",
            action="store_true",
            help="List cloud images only (default)",
        )
        list_parser.add_argument(
            "--platform",
            type=str,
            help="Filter local images by platform",
        )
        list_parser.add_argument(
            "--format",
            choices=["table", "json"],
            default="table",
            help="Output format for local images",
        )

        # push command
        push_parser = img_subparsers.add_parser(
            "push",
            help="Push image to cloud storage",
        )
        push_parser.add_argument(
            "name",
            help="Image name",
        )
        push_parser.add_argument(
            "--file",
            "-f",
            type=str,
            help="Path to image file (default: ~/.local/share/cua/images/<name>/data.img)",
        )
        push_parser.add_argument(
            "--tag",
            type=str,
            default="latest",
            help="Image tag (default: latest)",
        )
        push_parser.add_argument(
            "--type",
            type=str,
            default="qcow2",
            choices=["qcow2", "raw", "vmdk"],
            help="Image type (default: qcow2)",
        )

        # pull command
        pull_parser = img_subparsers.add_parser(
            "pull",
            help="Pull image from cloud storage",
        )
        pull_parser.add_argument(
            "name",
            help="Image name",
        )
        pull_parser.add_argument(
            "--tag",
            type=str,
            default="latest",
            help="Image tag (default: latest)",
        )
        pull_parser.add_argument(
            "--output",
            "-o",
            type=str,
            help="Output file path",
        )

        # delete command
        delete_parser = img_subparsers.add_parser(
            "delete",
            help="Delete an image",
        )
        delete_parser.add_argument(
            "name",
            help="Image name",
        )
        delete_parser.add_argument(
            "--tag",
            type=str,
            default="latest",
            help="Image tag (default: latest, for cloud images)",
        )
        delete_parser.add_argument(
            "--force",
            action="store_true",
            help="Skip confirmation",
        )
        delete_parser.add_argument(
            "--local",
            action="store_true",
            help="Delete a local image instead of cloud",
        )

        # create command (local)
        create_parser = img_subparsers.add_parser(
            "create",
            help="Create a local image from platform",
        )
        create_parser.add_argument(
            "platform",
            help="Platform name (e.g., linux-docker, windows-qemu)",
        )
        create_parser.add_argument("--name", help="Image name (default: same as platform)")
        create_parser.add_argument("--iso", help="Path to ISO file (for QEMU platforms)")
        create_parser.add_argument(
            "--download-iso",
            action="store_true",
            dest="download_iso",
            help="Download Windows 11 ISO (~6GB)",
        )
        create_parser.add_argument(
            "--docker-image", dest="docker_image", help="Override Docker image"
        )
        create_parser.add_argument(
            "--distro",
            default="ubuntu",
            choices=["ubuntu", "fedora"],
            help="Linux distribution",
        )
        create_parser.add_argument(
            "--version",
            default="14",
            help="OS version (e.g., 14 for Android, sonoma for macOS)",
        )
        create_parser.add_argument("--disk", default="64G", help="Disk size (default: 64G)")
        create_parser.add_argument("--memory", default="8G", help="Memory (default: 8G)")
        create_parser.add_argument("--cpus", default="8", help="CPU cores (default: 8)")
        create_parser.add_argument(
            "--winarena-apps",
            action="store_true",
            dest="winarena_apps",
            help="Install WinArena benchmark apps (Chrome, LibreOffice, VLC, etc.)",
        )
        create_parser.add_argument("--detach", "-d", action="store_true", help="Run in background")
        create_parser.add_argument("--force", action="store_true", help="Force recreation")
        create_parser.add_argument(
            "--skip-pull",
            action="store_true",
            dest="skip_pull",
            help="Don't pull Docker image",
        )
        create_parser.add_argument(
            "--no-kvm",
            action="store_true",
            dest="no_kvm",
            help="Disable KVM acceleration",
        )
        create_parser.add_argument(
            "--vnc-port",
            dest="vnc_port",
            help="VNC port (default: auto-allocate from 8006)",
        )
        create_parser.add_argument(
            "--api-port",
            dest="api_port",
            help="API port (default: auto-allocate from 5000)",
        )

        # info command (local)
        info_parser = img_subparsers.add_parser(
            "info",
            help="Show local image details",
        )
        info_parser.add_argument("name", help="Image name")

        # clone command (local)
        clone_parser = img_subparsers.add_parser(
            "clone",
            help="Clone a local image",
        )
        clone_parser.add_argument("source", help="Source image name")
        clone_parser.add_argument("target", help="Target image name")
        clone_parser.add_argument("--force", action="store_true", help="Overwrite if target exists")

        # shell command (local)
        shell_parser = img_subparsers.add_parser(
            "shell",
            help="Interactive shell into image (uses overlay by default)",
        )
        shell_parser.add_argument("name", help="Image name")
        shell_parser.add_argument(
            "--writable",
            action="store_true",
            help="Modify golden image directly (dangerous!)",
        )
        shell_parser.add_argument("--detach", "-d", action="store_true", help="Run in background")
        shell_parser.add_argument(
            "--vnc-port",
            dest="vnc_port",
            help="VNC port (default: auto-allocate from 8006)",
        )
        shell_parser.add_argument(
            "--api-port",
            dest="api_port",
            help="API port (default: auto-allocate from 5000)",
        )
        shell_parser.add_argument("--memory", default="8G", help="Memory (default: 8G)")
        shell_parser.add_argument("--cpus", default="8", help="CPU cores (default: 8)")
        shell_parser.add_argument(
            "--no-kvm",
            action="store_true",
            dest="no_kvm",
            help="Disable KVM acceleration",
        )


def execute(args: argparse.Namespace) -> int:
    """Execute image command based on subcommand."""
    from cua_cli.commands import local_image

    cmd = getattr(args, "image_command", None)

    if cmd in ("list", "ls"):
        return cmd_list(args)
    elif cmd == "push":
        return cmd_push(args)
    elif cmd == "pull":
        return cmd_pull(args)
    elif cmd == "delete":
        return cmd_delete(args)
    elif cmd == "create":
        return local_image.cmd_create(args)
    elif cmd == "info":
        return local_image.cmd_info(args)
    elif cmd == "clone":
        return local_image.cmd_clone(args)
    elif cmd == "shell":
        return local_image.cmd_shell(args)
    else:
        print_error("Usage: cua image <command>")
        print_info("Commands: list, push, pull, delete, create, info, clone, shell")
        return 1


def cmd_list(args: argparse.Namespace) -> int:
    """List images."""
    show_local = getattr(args, "local", False)
    show_cloud = getattr(args, "cloud", False)

    # If --local is set, delegate to local_image list
    if show_local and not show_cloud:
        from cua_cli.commands import local_image

        return local_image.cmd_local_list(args)

    # Default to cloud if neither specified, or if --cloud is set
    if not show_local:
        show_cloud = True

    all_images = []

    if show_cloud:
        cloud_images = run_async(_list_cloud_images())
        all_images.extend(cloud_images)

    if show_local:
        local_images = _list_local_images()
        all_images.extend(local_images)

    if getattr(args, "json", False):
        print_json(all_images)
        return 0

    if not all_images:
        print_info("No images found.")
        return 0

    columns = [
        ("name", "NAME"),
        ("type", "TYPE"),
        ("tag", "TAG"),
        ("size", "SIZE"),
        ("status", "STATUS"),
        ("created", "CREATED"),
    ]

    print_table(all_images, columns)
    return 0


async def _list_cloud_images() -> list[dict[str, Any]]:
    """List cloud images."""
    try:
        client = CloudAPIClient()
        images = await client.list_images()

        result = []
        for img in images:
            versions = img.get("versions", [])
            if versions:
                for ver in versions:
                    result.append(
                        {
                            "name": img.get("name", ""),
                            "type": img.get("image_type", ""),
                            "tag": ver.get("tag", ""),
                            "size": format_bytes(ver.get("size_bytes", 0)),
                            "status": ver.get("status", ""),
                            "created": _format_date(ver.get("created_at", "")),
                            "source": "cloud",
                        }
                    )
            else:
                result.append(
                    {
                        "name": img.get("name", ""),
                        "type": img.get("image_type", ""),
                        "tag": "-",
                        "size": "-",
                        "status": "-",
                        "created": _format_date(img.get("created_at", "")),
                        "source": "cloud",
                    }
                )
        return result
    except Exception as e:
        print_error(f"Failed to list cloud images: {e}")
        return []


def _list_local_images() -> list[dict[str, Any]]:
    """List local images."""
    result = []

    if not LOCAL_IMAGES_DIR.exists():
        return result

    for image_dir in LOCAL_IMAGES_DIR.iterdir():
        if not image_dir.is_dir():
            continue

        name = image_dir.name
        data_file = image_dir / "data.img"

        if data_file.exists():
            size = format_bytes(data_file.stat().st_size)
            created = datetime.fromtimestamp(data_file.stat().st_mtime).strftime("%Y-%m-%d")
        else:
            size = "-"
            created = "-"

        result.append(
            {
                "name": name,
                "type": "local",
                "tag": "latest",
                "size": size,
                "status": "ready" if data_file.exists() else "incomplete",
                "created": created,
                "source": "local",
            }
        )

    return result


def _format_date(date_str: str) -> str:
    """Format a date string."""
    if not date_str:
        return "-"
    try:
        dt = datetime.fromisoformat(date_str.replace("Z", "+00:00"))
        return dt.strftime("%Y-%m-%d")
    except Exception:
        return date_str[:10] if len(date_str) >= 10 else date_str


def cmd_push(args: argparse.Namespace) -> int:
    """Push an image to cloud storage."""
    name = args.name
    tag = args.tag
    image_type = args.type

    # Determine file path
    if args.file:
        file_path = Path(args.file)
    else:
        # Look in local images directory
        file_path = LOCAL_IMAGES_DIR / name / "data.img"

    if not file_path.exists():
        if args.file:
            print_error(f"File not found: {file_path}")
        else:
            print_error(f"Image not found: {name}")
            print_info(f"Looked in: {file_path}")
            print_info("Use --file to specify a custom path")
        return 1

    size_bytes = file_path.stat().st_size
    print_info(f"Pushing {file_path} ({format_bytes(size_bytes)})")

    return run_async(_push_image(name, tag, image_type, file_path, size_bytes))


async def _push_image(
    name: str, tag: str, image_type: str, file_path: Path, size_bytes: int
) -> int:
    """Push an image using multi-part upload."""
    from rich.progress import BarColumn, Progress, SpinnerColumn, TextColumn

    # Calculate checksum
    print_info("Calculating checksum...")
    checksum = calculate_file_hash(file_path)
    print_info(f"Checksum: {checksum}")

    client = CloudAPIClient()

    # Initiate upload
    print_info("Initiating upload...")
    status, data = await client.initiate_upload(
        name=name,
        tag=tag,
        image_type=image_type,
        size_bytes=size_bytes,
        checksum_sha256=checksum,
    )

    if status == 401:
        print_error("Unauthorized. Run 'cua auth login' again.")
        return 1
    if status == 409:
        print_error(f"Image version already exists: {name}:{tag}")
        return 1
    if status not in (200, 201):
        print_error(f"Failed to initiate upload: {data}")
        return 1

    upload_id = data.get("upload_id")
    part_size = data.get("part_size", DEFAULT_PART_SIZE)
    total_parts = data.get("total_parts", 1)

    print_info(f"Upload session: {upload_id}")
    print_info(f"Parts: {total_parts} x {format_bytes(part_size)}")

    # Read file into memory (for simplicity - could stream for very large files)
    file_data = file_path.read_bytes()

    completed_parts = []

    with Progress(
        SpinnerColumn(),
        TextColumn("[progress.description]{task.description}"),
        BarColumn(),
        TextColumn("[progress.percentage]{task.percentage:>3.0f}%"),
    ) as progress:
        task = progress.add_task("Uploading...", total=total_parts)

        for part_num in range(1, total_parts + 1):
            # Get signed URL
            status, url_data = await client.get_upload_part_url(name, upload_id, part_num)
            if status != 200:
                print_error(f"Failed to get upload URL for part {part_num}")
                await client.abort_upload(name, upload_id)
                return 1

            upload_url = url_data.get("upload_url")

            # Calculate part range
            start = (part_num - 1) * part_size
            end = min(start + part_size, size_bytes)
            part_data = file_data[start:end]

            # Upload part
            async with aiohttp.ClientSession() as session:
                async with session.put(
                    upload_url,
                    data=part_data,
                    headers={"Content-Type": "application/octet-stream"},
                ) as resp:
                    if resp.status not in (200, 201):
                        print_error(f"Failed to upload part {part_num}: {resp.status}")
                        await client.abort_upload(name, upload_id)
                        return 1

                    etag = resp.headers.get("ETag", "")
                    completed_parts.append({"part_number": part_num, "etag": etag})

            progress.update(task, advance=1)

    # Complete upload
    print_info("Completing upload...")
    status, result = await client.complete_upload(name, upload_id, completed_parts)

    if status not in (200, 201):
        print_error(f"Failed to complete upload: {result}")
        return 1

    print_success(f"Push complete: {name}:{tag}")
    if isinstance(result, dict):
        print_info(f"Version ID: {result.get('version_id', 'N/A')}")
        print_info(f"Status: {result.get('status', 'N/A')}")

    return 0


def cmd_pull(args: argparse.Namespace) -> int:
    """Pull an image from cloud storage."""
    name = args.name
    tag = args.tag
    output_path = args.output or f"{name}-{tag}.qcow2"

    print_info(f"Pulling {name}:{tag}...")
    return run_async(_pull_image(name, tag, Path(output_path)))


async def _pull_image(name: str, tag: str, output_path: Path) -> int:
    """Pull an image from cloud storage."""
    from rich.progress import (
        BarColumn,
        DownloadColumn,
        Progress,
        SpinnerColumn,
        TextColumn,
    )

    client = CloudAPIClient()

    # Get download URL
    status, data = await client.get_download_url(name, tag)

    if status == 401:
        print_error("Unauthorized. Run 'cua auth login' again.")
        return 1
    if status == 404:
        print_error(f"Image not found: {name}:{tag}")
        return 1
    if status != 200:
        print_error(f"Failed to get download URL: {data}")
        return 1

    download_url = data.get("download_url")
    size_bytes = data.get("size_bytes", 0)
    expected_checksum = data.get("checksum_sha256", "")

    print_info(f"Size: {format_bytes(size_bytes)}")
    print_info(f"Downloading to {output_path}...")

    # Download file
    async with aiohttp.ClientSession() as session:
        async with session.get(download_url) as resp:
            if resp.status != 200:
                print_error(f"Download failed: {resp.status}")
                return 1

            with Progress(
                SpinnerColumn(),
                TextColumn("[progress.description]{task.description}"),
                BarColumn(),
                DownloadColumn(),
            ) as progress:
                task = progress.add_task("Downloading...", total=size_bytes)

                with open(output_path, "wb") as f:
                    async for chunk in resp.content.iter_chunked(8192):
                        f.write(chunk)
                        progress.update(task, advance=len(chunk))

    # Verify checksum
    if expected_checksum:
        print_info("Verifying checksum...")
        downloaded_checksum = calculate_file_hash(output_path)

        if downloaded_checksum != expected_checksum:
            print_error("Checksum mismatch! Download may be corrupted.")
            print_error(f"Expected: {expected_checksum}")
            print_error(f"Got: {downloaded_checksum}")
            return 1

        print_success(f"Checksum verified: {downloaded_checksum}")

    print_success(f"Pull complete: {output_path}")
    return 0


def cmd_delete(args: argparse.Namespace) -> int:
    """Delete an image."""
    # If --local flag, delegate to local_image delete
    if getattr(args, "local", False):
        from cua_cli.commands import local_image

        return local_image.cmd_local_delete(args)

    # Cloud delete
    name = args.name
    tag = args.tag

    if not args.force:
        print_info(f"This will delete {name}:{tag}. Use --force to confirm.")
        return 1

    print_info(f"Deleting {name}:{tag}...")
    return run_async(_delete_image(name, tag))


async def _delete_image(name: str, tag: str) -> int:
    """Delete an image from cloud storage."""
    client = CloudAPIClient()

    status, data = await client.delete_image(name, tag)

    if status == 401:
        print_error("Unauthorized. Run 'cua auth login' again.")
        return 1
    if status == 404:
        print_error(f"Image not found: {name}:{tag}")
        return 1
    if status not in (200, 202, 204):
        print_error(f"Delete failed: {data}")
        return 1

    print_success(f"Deleted: {name}:{tag}")
    return 0

File: libs/typescript/cua-cli/src/commands/image.ts
import type { Argv } from 'yargs';
import { createHash } from 'crypto';
import { ensureApiKeyInteractive } from '../auth';
import { http } from '../http';
import { clearApiKey } from '../storage';

// Types for image API responses
type ImageInfo = {
  image_id: number;
  name: string;
  description: string;
  image_type: string;
  created_at: string;
  versions: ImageVersionInfo[];
};

type ImageVersionInfo = {
  version_id: number;
  tag: string;
  size_bytes: number;
  checksum_sha256: string;
  status: string;
  created_at: string;
};

type UploadSession = {
  upload_id: string;
  version_id: number;
  part_size: number;
  total_parts: number;
  expires_at: string;
};

type PartInfo = {
  part_number: number;
  etag: string;
};

// Default part size: 100MB
const DEFAULT_PART_SIZE = 100 * 1024 * 1024;

// Helper to format bytes
function formatBytes(bytes: number): string {
  if (bytes === 0) return '0 B';
  const k = 1024;
  const sizes = ['B', 'KB', 'MB', 'GB', 'TB'];
  const i = Math.floor(Math.log(bytes) / Math.log(k));
  return parseFloat((bytes / Math.pow(k, i)).toFixed(2)) + ' ' + sizes[i];
}

// Helper to calculate SHA256 hash of a file
async function calculateFileHash(filePath: string): Promise<string> {
  const file = Bun.file(filePath);
  const buffer = await file.arrayBuffer();
  const hash = createHash('sha256');
  hash.update(Buffer.from(buffer));
  return hash.digest('hex');
}

// Helper for printing tables
function printTable(headers: string[], rows: string[][]) {
  const allRows = [headers, ...rows];
  const widths: number[] = new Array(headers.length).fill(0);

  for (const r of allRows)
    for (let i = 0; i < headers.length; i++)
      widths[i] = Math.max(widths[i] ?? 0, (r[i] ?? '').length);

  for (const r of allRows)
    console.log(r.map((c, i) => (c ?? '').padEnd(widths[i] ?? 0)).join('  '));
}

// Command handlers
const listHandler = async (argv: Record<string, unknown>) => {
  const token = await ensureApiKeyInteractive();
  const res = await http('/v1/images', { token });

  if (res.status === 401) {
    clearApiKey();
    console.error("Unauthorized. Try 'cua login' again.");
    process.exit(1);
  }
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    console.error(`Request failed: ${res.status}`, (body as any).error || '');
    process.exit(1);
  }

  const images = (await res.json()) as ImageInfo[];

  if (argv.json) {
    console.log(JSON.stringify(images, null, 2));
    return;
  }

  if (images.length === 0) {
    console.log('No images found');
    return;
  }

  const headers = ['NAME', 'TYPE', 'TAG', 'SIZE', 'STATUS', 'CREATED'];
  const rows: string[][] = [];

  for (const img of images) {
    if (img.versions && img.versions.length > 0) {
      for (const ver of img.versions) {
        rows.push([
          img.name,
          img.image_type,
          ver.tag,
          formatBytes(ver.size_bytes),
          ver.status,
          new Date(ver.created_at).toLocaleDateString(),
        ]);
      }
    } else {
      rows.push([
        img.name,
        img.image_type,
        '-',
        '-',
        '-',
        new Date(img.created_at).toLocaleDateString(),
      ]);
    }
  }

  printTable(headers, rows);
};

const pushHandler = async (argv: Record<string, unknown>) => {
  const token = await ensureApiKeyInteractive();
  const name = String(argv.name);
  const tag = String(argv.tag || 'latest');
  const imageType = String(argv.type || 'qcow2');

  // Determine file path - use --file if provided, otherwise look in cua-bench images directory
  let filePath: string;
  if (argv.file) {
    filePath = String(argv.file);
  } else {
    // Look for image in cua-bench local storage
    const home = process.env.HOME || process.env.USERPROFILE || '';
    const cuaBenchPath = `${home}/.local/share/cua-bench/images/${name}/data.img`;
    filePath = cuaBenchPath;
  }

  // Check if file exists and get size
  const file = Bun.file(filePath);
  const exists = await file.exists();
  if (!exists) {
    if (argv.file) {
      console.error(`File not found: ${filePath}`);
    } else {
      console.error(`Image not found: ${name}`);
      console.error(`Looked in: ${filePath}`);
      console.error(`Use --file to specify a custom path`);
    }
    process.exit(1);
  }

  const sizeBytes = file.size;
  console.log(`Pushing ${filePath} (${formatBytes(sizeBytes)})`);

  // Calculate checksum
  console.log('Calculating checksum...');
  const checksum = await calculateFileHash(filePath);
  console.log(`Checksum: ${checksum}`);

  // Initiate upload
  console.log('Initiating upload...');
  const initRes = await http(`/v1/images/${encodeURIComponent(name)}/upload`, {
    token,
    method: 'POST',
    body: {
      tag,
      image_type: imageType,
      size_bytes: sizeBytes,
      checksum_sha256: checksum,
    },
  });

  if (initRes.status === 401) {
    clearApiKey();
    console.error("Unauthorized. Try 'cua login' again.");
    process.exit(1);
  }
  if (initRes.status === 409) {
    console.error(`Image version already exists: ${name}:${tag}`);
    process.exit(1);
  }
  if (!initRes.ok) {
    const body = await initRes.json().catch(() => ({}));
    console.error(
      `Failed to initiate upload: ${initRes.status}`,
      (body as any).error || ''
    );
    process.exit(1);
  }

  const session = (await initRes.json()) as UploadSession;
  console.log(`Upload session: ${session.upload_id}`);
  console.log(
    `Parts: ${session.total_parts} x ${formatBytes(session.part_size)}`
  );

  // Upload parts
  const completedParts: PartInfo[] = [];
  const fileBuffer = await file.arrayBuffer();

  for (let partNum = 1; partNum <= session.total_parts; partNum++) {
    const start = (partNum - 1) * session.part_size;
    const end = Math.min(start + session.part_size, sizeBytes);
    const partData = fileBuffer.slice(start, end);

    // Get signed URL for this part
    const urlRes = await http(
      `/v1/images/${encodeURIComponent(name)}/upload/${session.upload_id}/part/${partNum}`,
      { token }
    );

    if (!urlRes.ok) {
      console.error(`Failed to get upload URL for part ${partNum}`);
      // Abort upload
      await http(
        `/v1/images/${encodeURIComponent(name)}/upload/${session.upload_id}`,
        {
          token,
          method: 'DELETE',
        }
      );
      process.exit(1);
    }

    const urlData = (await urlRes.json()) as {
      upload_url: string;
      expires_in: number;
    };

    // Upload the part
    process.stdout.write(
      `\rUploading part ${partNum}/${session.total_parts}...`
    );

    const uploadRes = await fetch(urlData.upload_url, {
      method: 'PUT',
      body: partData,
      headers: {
        'Content-Type': 'application/octet-stream',
      },
    });

    if (!uploadRes.ok) {
      console.error(`\nFailed to upload part ${partNum}: ${uploadRes.status}`);
      // Abort upload
      await http(
        `/v1/images/${encodeURIComponent(name)}/upload/${session.upload_id}`,
        {
          token,
          method: 'DELETE',
        }
      );
      process.exit(1);
    }

    // Get ETag from response
    const etag = uploadRes.headers.get('ETag') || '';
    completedParts.push({ part_number: partNum, etag });
  }

  console.log('\nCompleting upload...');

  // Complete the upload
  const completeRes = await http(
    `/v1/images/${encodeURIComponent(name)}/upload/${session.upload_id}/complete`,
    {
      token,
      method: 'POST',
      body: { parts: completedParts },
    }
  );

  if (!completeRes.ok) {
    const body = await completeRes.json().catch(() => ({}));
    console.error(
      `Failed to complete upload: ${completeRes.status}`,
      (body as any).error || ''
    );
    process.exit(1);
  }

  const result = (await completeRes.json()) as ImageVersionInfo;
  console.log(`Push complete: ${name}:${tag}`);
  console.log(`Version ID: ${result.version_id}`);
  console.log(`Status: ${result.status}`);
};

const pullHandler = async (argv: Record<string, unknown>) => {
  const token = await ensureApiKeyInteractive();
  const name = String(argv.name);
  const tag = String(argv.tag || 'latest');
  const outputPath = String(argv.output || `${name}-${tag}.qcow2`);

  console.log(`Pulling ${name}:${tag}...`);

  // Get download URL
  const urlRes = await http(
    `/v1/images/${encodeURIComponent(name)}/download?tag=${encodeURIComponent(tag)}`,
    {
      token,
    }
  );

  if (urlRes.status === 401) {
    clearApiKey();
    console.error("Unauthorized. Try 'cua login' again.");
    process.exit(1);
  }
  if (urlRes.status === 404) {
    console.error(`Image not found: ${name}:${tag}`);
    process.exit(1);
  }
  if (!urlRes.ok) {
    const body = await urlRes.json().catch(() => ({}));
    console.error(
      `Failed to get download URL: ${urlRes.status}`,
      (body as any).error || ''
    );
    process.exit(1);
  }

  const urlData = (await urlRes.json()) as {
    download_url: string;
    size_bytes: number;
    checksum_sha256: string;
  };
  console.log(`Size: ${formatBytes(urlData.size_bytes)}`);

  // Download the file
  console.log(`Downloading to ${outputPath}...`);
  const downloadRes = await fetch(urlData.download_url);

  if (!downloadRes.ok) {
    console.error(`Download failed: ${downloadRes.status}`);
    process.exit(1);
  }

  const buffer = await downloadRes.arrayBuffer();
  await Bun.write(outputPath, buffer);

  // Verify checksum
  console.log('Verifying checksum...');
  const downloadedChecksum = await calculateFileHash(outputPath);

  if (downloadedChecksum !== urlData.checksum_sha256) {
    console.error('Checksum mismatch! Download may be corrupted.');
    console.error(`Expected: ${urlData.checksum_sha256}`);
    console.error(`Got: ${downloadedChecksum}`);
    process.exit(1);
  }

  console.log(`Pull complete: ${outputPath}`);
  console.log(`Checksum verified: ${downloadedChecksum}`);
};

const deleteHandler = async (argv: Record<string, unknown>) => {
  const token = await ensureApiKeyInteractive();
  const name = String(argv.name);
  const tag = String(argv.tag || 'latest');

  if (!argv.force) {
    console.log(`This will delete ${name}:${tag}. Use --force to confirm.`);
    process.exit(1);
  }

  console.log(`Deleting ${name}:${tag}...`);

  const res = await http(
    `/v1/images/${encodeURIComponent(name)}?tag=${encodeURIComponent(tag)}`,
    {
      token,
      method: 'DELETE',
    }
  );

  if (res.status === 401) {
    clearApiKey();
    console.error("Unauthorized. Try 'cua login' again.");
    process.exit(1);
  }
  if (res.status === 404) {
    console.error(`Image not found: ${name}:${tag}`);
    process.exit(1);
  }
  if (!res.ok) {
    const body = await res.json().catch(() => ({}));
    console.error(`Delete failed: ${res.status}`, (body as any).error || '');
    process.exit(1);
  }

  console.log(`Deleted: ${name}:${tag}`);
};

// Register commands
export function registerImageCommands(y: Argv) {
  y.command(
    ['image', 'img'],
    'Manage VM images (push, pull, delete)',
    (y) => {
      return y
        .command(
          ['list', 'ls'],
          'List all images in your workspace',
          (y) =>
            y.option('json', {
              type: 'boolean',
              default: false,
              describe: 'Output in JSON format',
            }),
          listHandler
        )
        .command(
          'push <name>',
          'Push a VM image to cloud storage',
          (y) =>
            y
              .positional('name', {
                type: 'string',
                describe: 'Image name',
                demandOption: true,
              })
              .option('file', {
                alias: 'f',
                type: 'string',
                describe:
                  'Path to image file (defaults to ~/.local/share/cua-bench/images/<name>/data.img)',
              })
              .option('tag', {
                type: 'string',
                default: 'latest',
                describe: 'Image tag/version',
              })
              .option('type', {
                type: 'string',
                default: 'qcow2',
                choices: ['qcow2', 'raw', 'vmdk'],
                describe: 'Image type',
              }),
          pushHandler
        )
        .command(
          'pull <name>',
          'Pull a VM image from cloud storage',
          (y) =>
            y
              .positional('name', {
                type: 'string',
                describe: 'Image name',
                demandOption: true,
              })
              .option('tag', {
                type: 'string',
                default: 'latest',
                describe: 'Image tag/version',
              })
              .option('output', {
                alias: 'o',
                type: 'string',
                describe: 'Output file path',
              }),
          pullHandler
        )
        .command(
          'delete <name>',
          'Delete an image version from cloud storage',
          (y) =>
            y
              .positional('name', {
                type: 'string',
                describe: 'Image name',
                demandOption: true,
              })
              .option('tag', {
                type: 'string',
                default: 'latest',
                describe: 'Image tag/version',
              })
              .option('force', {
                alias: 'f',
                type: 'boolean',
                default: false,
                describe: 'Skip confirmation',
              }),
          deleteHandler
        )
        .demandCommand(1, 'You must provide an image command');
    },
    () => {}
  );

  return y;
}
